examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,2012 @@
+                # pretrained feature network for the discriminator operates in pixel space rather than latent space.
+                unscaled_student_x_0 = (1 / vae.config.scaling_factor) * student_x_0
+                # Perform batched decode with batch size of at most args.vae_encode_batch_size
+                student_gen_image = []
+                for i in range(0, unscaled_student_x_0.shape[0], args.vae_encode_batch_size):
+                    student_gen_image.append(
+                        vae.decode(
+                            unscaled_student_x_0[i : i + args.vae_encode_batch_size].to(dtype=weight_dtype)
+                        ).sample
+                    )
+                student_gen_image = torch.cat(student_gen_image, dim=0)
+
+                # 2. Get discriminator real/fake outputs on the real and fake (generated) images respectively.
+                disc_output_real = discriminator(pixel_values.float(), text_embedding, image_embedding)
+                disc_output_fake = discriminator(student_gen_image.detach().float(), text_embedding, image_embedding)
+
+                # 3. Calculate the discriminator real adversarial loss terms.
+                d_logits_real = disc_output_real.logits
+                # Use hinge loss (see section 3.2, Equation 3 of paper)
+                d_adv_loss_real = torch.mean(F.relu(torch.ones_like(d_logits_real) - d_logits_real))
+
+                # 4. Calculate the discriminator R1 gradient penalty term with respect to the gradients from the real
+                # data.
+                d_r1_regularizer = 0
+                for k, head in discriminator.heads.items():
+                    head_grad_params = torch.autograd.grad(
+                        outputs=d_adv_loss_real, inputs=head.parameters(), create_graph=True
+                    )
+                    head_grad_norm = 0
+                    for grad in head_grad_params:
+                        head_grad_norm += grad.pow(2).sum()
+                    head_grad_norm = head_grad_norm.sqrt()
+                    d_r1_regularizer += head_grad_norm
+
+                d_loss_real = d_adv_loss_real + args.discriminator_r1_strength * d_r1_regularizer
+                accelerator.backward(d_loss_real, retain_graph=True)
+
+                # 5. Calculate the discriminator fake adversarial loss terms.
+                d_logits_fake = disc_output_fake.logits
+                # Use hinge loss (see section 3.2, Equation 3 of paper)
+                d_adv_loss_fake = torch.mean(F.relu(torch.ones_like(d_logits_fake) + d_logits_fake))
+                accelerator.backward(d_adv_loss_fake)
+
+                d_total_loss = d_loss_real + d_adv_loss_fake
+
+                if accelerator.sync_gradients:
+                    accelerator.clip_grad_norm_(discriminator.parameters(), args.max_grad_norm)
+                discriminator_optimizer.step()
```

Comment:
if i understand correct, when the `gradient_accumulation_steps > 1`current implementation seems result in a simultaneous gradient descent up to the last accumulation step for each batch, when `gradient_accumulation_steps == 1` it is alternating gradient descent, while in the stylegan-t it is always alternating gradient descent

Reply:
How important is that to follow? Is it absolutely a must for training stability?

Reply:
convergence behavior of simultaneous gradient descent and alternating gradient descent are different when achieving min-max equilibrium, from those gan implementations lately, seems alternating gradient descent is usually adopted, not sure how important they could help for the case of ADD though